algebra-core/src/fields/models/fp4.rs

```diff
@@ -0,0 +1,427 @@
+use crate::{
+    io::{Read, Result as IoResult, Write},
+    BigInteger, CanonicalDeserialize, CanonicalDeserializeWithFlags, CanonicalSerialize,
+    CanonicalSerializeWithFlags, ConstantSerializedSize, EmptyFlags, Flags, SerializationError,
+    UniformRand,
+};
+use core::{
+    cmp::{Ord, Ordering, PartialOrd},
+    fmt,
+    marker::PhantomData,
+    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+};
+use num_traits::{One, Zero};
+use rand::{
+    distributions::{Distribution, Standard},
+    Rng,
+};
+
+use crate::{
+    bytes::{FromBytes, ToBytes},
+    fields::{Field, Fp2, Fp2Parameters},
+};
+
+pub trait Fp4Parameters: 'static + Send + Sync {
+    type Fp2Params: Fp2Parameters;
+
+    const NONRESIDUE: <Self::Fp2Params as Fp2Parameters>::Fp;
+
+    /// Coefficients for the Frobenius automorphism.
+    /// non_residue^((modulus^i-1)/4) for i=0,1,2,3
+    const FROBENIUS_COEFF_FP4_C1: [<Self::Fp2Params as Fp2Parameters>::Fp; 4];
+
+    #[inline(always)]
+    fn mul_fp_by_nonresidue(fe: &Fp2<Self::Fp2Params>) -> Fp2<Self::Fp2Params> {
```

Comment:
Should this be `mul_fp2_by_nonresidue`?

```suggestion
    fn mul_fp2_by_nonresidue(fe: &Fp2<Self::Fp2Params>) -> Fp2<Self::Fp2Params> {
```